README.md
![Logo](images/logo.jpg?raw=true "Carto plugin for QGis")
================================================
# Carto plugin for QGis
### The best of your Desktop GIS with the best of the web GIS.

![QGisCartoDB](https://lh3.googleusercontent.com/aVV2C4SigQ6Tu15LCO9kSeHNDZIQe3qr1alt4NMMq6Y=w1161-h659-no "QGis Carto")

## Features

* Manage Carto connections.
* Add Carto layers to QGIS projects.
* Create QGIS layers from SQL Queries.
* Add features to Carto tables.
* Edit data and update Carto tables.
* Edit data and update geometries.
* Delete features.
* Upload vector layers to Carto.
* Create Carto visualizations using QGIS

## Documentation

* [Carto plugin for QGIS Wiki](https://github.com/gkudos/qgis-cartodb/wiki)


## Sponsors

This project was sponsored by:

### Carto

[![Carto](images/logo_carto.png)](https://carto.com/)

Versions 0.1.x / 0.2


### Kudos Ltda.

[![Kudos Ltda](https://lh3.googleusercontent.com/63fyD2eSIM58HPj_rZfyjyVryahxPinVWWdzL7OvLvg=w640-h261-no)](http://gkudos.com/)


CHANGELOG.md
<a name="0.2.7"></a>
### 0.2.7 (2017-06-06)

#### Fixes

* Fix #51 CARTO rebranding applied in QGIS UI.

<a name="0.2.6"></a>
### 0.2.6 (2016-09-16)

#### Fixes

* Fix #47 changing entry points to carto.com domain.
* Resolve some deprecated warnings.

<a name="0.2.5"></a>
### 0.2.5 (2016-07-08)

#### Fixes

* Fix #44 problems with domain changes.

### 0.2.4 (2016-03-14)

#### Refactoring

* Change 'Add CartoDB Layer' button to layer toolbar on QGIS >= 2.12. (Left side on QGIS).
* Change 'Add SQL CartoDB Layer' button to layer toolbar on QGIS >= 2.12. (Left side on QGIS).

#### Fixes

* Fix #40 error when the layer name contains non-ascii characters

<a name="0.2.3"></a>
### 0.2.3 (2015-10-04)

#### Fixes

* Fix #33 Error updating recently uploaded data

<a name="0.2.2"></a>
### 0.2.2 (2015-09-02)

#### Enhancements

* Change privacy for maps (public checkbox).
* Change some texts.

<a name="0.2.1"></a>
### 0.2.1 (2015-08-19)

#### Enhancements

* Create maps without select layers.
* Add status to upload items.
* Change icons.

#### Fixes

* Fix order layer in maps.
* Fix #30 When you push a new layer to CartoDB, it doesn't register the layer with your current project as being a CartoDB layer.

<a name="0.2.0"></a>
### 0.2.0 (2015-08-10)

#### Enhancements

* Add functionality to Create Map buttons.

#### Fixes

* Fix #25 Error creating map from CartoDB.
* Fix #5 Filter data by attribute or extent.
* Fix problems with SQL layers in maps.
* Fix #31 Style Categorized: The generation of CartoCSS fails if there is data with special characters.
* Remove unnecessary files in the plugin package.

<a name="0.1.9"></a>
### 0.1.9 (2015-07-22)

#### Enhancements

* Default symbology for not supported symbols.
* Add messages for not supported symbols.
* Add uploaded message.
* Change icons.

#### Fixes

* Fix #29 Error updating layer single user account.
* Fix #27 Error getting columns in SQL dialog.


<a name="0.1.8"></a>
### 0.1.8 (2015-07-11)

#### Enhancements

* Remove connection controls in SQL Editor.
* Add composite mode in create maps.
* Add dash style to lines in create maps.
* Add join style to lines in create maps.

<a name="0.1.7"></a>
### 0.1.7 (2015-07-10)

#### Fixes

* Fix #19 Do not allow to edit read-only layers.
* Show owner in shared layers, issue #18.
* Fix #26 Encoding error creating map.


<a name="0.1.6"></a>
### 0.1.6 (2015-06-20)

#### Enhancements

* Remove scroll in user table list.
* Ascending order in user table list.
* Download datasets in spatialite format.

#### Fixes

* Fix #23 Error conecting multiuser account.

<a name="0.1.5"></a>
### 0.1.5 (2015-06-16)

#### New Features

* Upload layers to CartoDB (More formats).
* Create basic visualization. Support:
  * Borders.
  * Fill.
  * Label.

<a name="0.1.4"></a>
### 0.1.4 (2015-06-05)

#### New Features

* Upload layers to CartoDB.

#### Translate

* Add spanish translation.

#### Fixes

* Fix avatar size.
* Fix #21 Insert data without complete fields.
* Enable buttons only if there is at least one created connection.

<a name="0.1.3"></a>
### 0.1.3 (2015-05-24)

#### New Features

* Filter tables by current extent.
* Add more info for table list.
* Use threads for download data.

#### Fixes.

* Fix error at multiusers accounts #15, again.

<a name="0.1.2"></a>
### 0.1.2 (2015-05-07)

#### New Features

* Filter tables by name. Fix #4
* Indicate READ_ONLY tables for multiuser accounts.
* Add unit test to project.
* Add user information at main dialog.
* Load Tables from viz API.
* Paginate tables on scroll.

#### Fixes.

* Fix error at multiusers accounts #15

<a name="0.1.1"></a>
### 0.1.1 (2014-11-27)

* Fix error when repeat layer name at spatialite database.
#### New Features

* Load cartodb layers from SQL Queries.

#### Fixes.

* New connection dialog is now a modal window.

<a name="0.1.0"></a>
### 0.1.0 (2014-09-23)

#### Features

* Add, edit and delete features.
* Add, edit and delete data.
* Manage connection.


dialogs/Main.py
"""
/***************************************************************************
CartoDB Plugin
A QGIS plugin

----------------------------------------------------------------------------
begin                : 2014-09-08
copyright            : (C) 2015 by Michael Salgado, Kudos Ltda.
email                : michaelsalgado@gkudos.com, info@gkudos.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""
from PyQt4.QtCore import QUrl, QEventLoop, pyqtSignal, pyqtSlot, Qt, qDebug
from PyQt4.QtGui import QApplication, QDialog, QMessageBox, QListWidgetItem, QIcon

from qgis.core import QgsMessageLog

from QgisCartoDB.cartodb import CartoDBAPIKey, CartoDBException, CartoDBApi
from QgisCartoDB.dialogs.UserData import CartoDBUserDataDialog
from QgisCartoDB.ui.UI_CartoDBPlugin import Ui_CartoDBPlugin
from QgisCartoDB.utils import CartoDBPluginWorker
from QgisCartoDB.widgets import CartoDBDatasetsListItem

import QgisCartoDB.resources

import copy
import json


# Create the dialog for CartoDBPlugin
class CartoDBPluginDialog(CartoDBUserDataDialog):
    def __init__(self, toolbar, parent=None):
        CartoDBUserDataDialog.__init__(self, toolbar, parent)

        # Set up the user interface from Designer.
        self.ui = Ui_CartoDBPlugin()
        self.ui.setupUi(self)
        self.ui.searchTX.textChanged.connect(self.filterTables)
        self.ui.sqlBT.clicked.connect(self.onSQL)
        # self.ui.tablesList.verticalScrollBar().valueChanged.connect(self.onScroll)

        self.isLoadingTables = False
        self.noLoadTables = False

        self.initUserConnection()
        self.loadSQL = False

    def getTablesListSelectedItems(self):
        return self.ui.tablesList.selectedItems()

    def getItemWidget(self, item):
        return self.ui.tablesList.itemWidget(item)

    def filterByExtent(self):
        return self.ui.extentCH.isChecked()

    def onSQL(self):
        self.loadSQL = True
        self.reject()

    @pyqtSlot()
    def connect(self):
        # Get tables from CartoDB.
        self.tablesPage = 1
        self.noLoadTables = False
        self.ui.searchTX.setText('')
        self.getTables(self.currentUser, self.currentApiKey, self.currentMultiuser, self.currentHostName)

    def filterTables(self):
        text = self.ui.searchTX.text()
        if text == '':
            newVisualizations = self.visualizations
        else:
            newVisualizations = [t for t in self.visualizations if text in t['name']]
        self.updateList(newVisualizations)

    def updateList(self, visualizations):
        self.ui.tablesList.clear()
        for visualization in visualizations:
            item = QListWidgetItem(self.ui.tablesList)

            owner = None
            owner = visualization['permission']['owner']['username']

            widget = CartoDBDatasetsListItem(
                visualization['name'], owner, visualization['table']['size'], visualization['table']['row_count'],
                shared=(owner != self.currentUser), multiuser=self.currentMultiuser)
            # item.setText(visualization['name'])
            readonly = False
            # qDebug('Vis:' + json.dumps(visualization, sort_keys=True, indent=2, separators=(',', ': ')))
            if visualization['permission'] is not None and owner != self.currentUser and \
               visualization['permission']['acl'] is not None:
                for acl in visualization['permission']['acl']:
                    if acl['type'] == 'user' and 'username' in acl['entity'] and acl['entity']['username'] == self.currentUser and \
                       acl['access'] == 'r':
                        readonly = True
                        break

            widget.readonly = readonly
            if readonly:
                widget.setTextColor('#999999')

            item.setSizeHint(widget.sizeHint())
            # item.setIcon(QIcon(":/plugins/qgis-cartodb/images/icons/layers.png"))
            self.ui.tablesList.setItemWidget(item, widget)

    def getTables(self, cartodbUser, apiKey, multiuser=False):
        cartoDBApi = CartoDBApi(cartodbUser, apiKey, multiuser)
        cartoDBApi.fetchContent.connect(self.cbTables)
        cartoDBApi.error.connect(self.error)
        self.isLoadingTables = True
        # cartoDBApi.getUserTables(self.tablesPage)
        cartoDBApi.getUserTables(1, 100000)

    @pyqtSlot(dict)
    def cbTables(self, data):
        if 'error' in data:
            return

        self.totalTables = data['total_user_entries']
        self.totalShared = data['total_shared']

        if len(data['visualizations']) == 0:
            self.noLoadTables = True

        if self.tablesPage == 1:
            self.visualizations = data['visualizations']
        else:
            self.visualizations.extend(data['visualizations'])

        self.visualizations.reverse()

        self.updateList(self.visualizations)
        self.ui.searchTX.setEnabled(True)
        self.isLoadingTables = False
        self.setUpUserData()

    def onScroll(self, val):
        maximum = self.ui.tablesList.verticalScrollBar().maximum()
        if val >= maximum - 4 and not self.isLoadingTables and not self.noLoadTables:
            self.tablesPage = self.tablesPage + 1
            self.getTables(self.currentUser, self.currentApiKey, self.currentMultiuser)

    def showEvent(self, event):
        worker = CartoDBPluginWorker(self, 'connect')
        worker.start()

    def error(self):
        QMessageBox.warning(self, QApplication.translate('CartoDBPlugin', 'Error'),
                            QApplication.translate('CartoDBPlugin', 'Error loading data from CartoDB'))
